subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1372 @@
+		return;
+	}
+
+	if (data->if_state == IF_STATE_SPEED_CHANGE_SENT) {
+		data->if_state = IF_STATE_DONE;
+	}
+}
+
+static void usbd_cdc_ncm_update(struct usbd_class_data *const c_data,
+				const uint8_t iface, const uint8_t alternate)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+	uint8_t data_iface = desc->if1_1.bInterfaceNumber;
+	int ret;
+
+	LOG_WRN("New configuration, interface %u alternate %u",
+		iface, alternate);
+
+	if (data_iface == iface && alternate == 0) {
+		net_if_carrier_off(data->iface);
+
+		LOG_DBG("Skip iface %u alternate %u", iface, alternate);
+
+		data->tx_seq = 0;
+	}
+
+	if (data_iface == iface && alternate == 1) {
+		net_if_carrier_on(data->iface);
+
+		ret = cdc_ncm_out_start(c_data);
+		if (ret < 0) {
+			LOG_ERR("Failed to start OUT transfer (%d)", ret);
+		}
+	}
+}
+
+static void usbd_cdc_ncm_enable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+
+	atomic_set_bit(&data->state, CDC_NCM_CLASS_ENABLED);
+
+	LOG_DBG("Configuration enabled");
+}
+
+static void usbd_cdc_ncm_disable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+
+	if (atomic_test_and_clear_bit(&data->state, CDC_NCM_CLASS_ENABLED)) {
+		net_if_carrier_off(data->iface);
+	}
+
+	atomic_clear_bit(&data->state, CDC_NCM_CLASS_SUSPENDED);
+
+	LOG_DBG("Configuration disabled");
+}
+
+static void usbd_cdc_ncm_suspended(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+
+	atomic_set_bit(&data->state, CDC_NCM_CLASS_SUSPENDED);
+}
+
+static void usbd_cdc_ncm_resumed(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+
+	atomic_clear_bit(&data->state, CDC_NCM_CLASS_SUSPENDED);
+}
+
+static int usbd_cdc_ncm_ctd(struct usbd_class_data *const c_data,
+			    const struct usb_setup_packet *const setup,
+			    const struct net_buf *const buf)
+{
+	if (setup->RequestType.recipient == USB_REQTYPE_RECIPIENT_INTERFACE) {
+		if (setup->bRequest == SET_ETHERNET_PACKET_FILTER) {
+			LOG_DBG("bRequest 0x%02x (%s) not implemented",
+				setup->bRequest, "SetPacketFilter");
+			return 0;
+		}
+
+		if (setup->bRequest == SET_NTB_INPUT_SIZE) {
+			LOG_DBG("bRequest 0x%02x (%s) not implemented",
+				setup->bRequest, "SetNtbInputSize");
+			return 0;
+		}
+
+		if (setup->bRequest == SET_NTB_FORMAT) {
+			LOG_DBG("bRequest 0x%02x (%s) not implemented",
+				setup->bRequest, "SetNtbFormat");
+			return 0;
+		}
+	}
+
+	LOG_DBG("bmRequestType 0x%02x bRequest 0x%02x unsupported",
+		setup->bmRequestType, setup->bRequest);
+	errno = -ENOTSUP;
+
+	return -1;
```

Comment:
It should be `return 0;`. Setting errno and `return 0;` means protocol error, but returning -1 would mean an unrecoverable error and would be passed to the user application.